feat: use new begin and end block events

guoguojin opened the pull request:
Closes #6211 
Closes #6414

guoguojin committed:
feat: use new begin and end block events

guoguojin committed:
chore: update change log

guoguojin committed:
chore: address linting errors

pscott31 commented on datanode/sqlstore/node.go:
```diff
@@ -43,6 +44,7 @@ func NewNode(connectionSource *ConnectionSource) *Node {
 
 func (store *Node) UpsertNode(ctx context.Context, node *entities.Node) error {
 	defer metrics.StartSQLQuery("Node", "UpsertNode")()
+	store.log.Info("Upserting node", zap.String("node", node.ID.String()))
```
super minor but I think the convention here is to use logging.String()

guoguojin replied:
Sorry, that should have been removed. I will update it.

pscott31 commented on core/processor/abci.go:
```diff
@@ -625,15 +625,29 @@ func (app *App) OnInitChain(req tmtypes.RequestInitChain) tmtypes.ResponseInitCh
 	// let's assume genesis block is block 0
```
Is this comment still true?

guoguojin committed:
chore: address review comments

guoguojin committed:
Merge branch 'develop' into feat/6211-handle-block-start-end-events

ettec commented on datanode/broker/sqlstore_broker.go:
```diff
@@ -130,33 +130,35 @@ func (b *SQLStoreBroker) waitForFirstBlock(ctx context.Context, errCh <-chan err
 	} else if errors.Is(err, sqlstore.ErrNoLastBlock) {
 		lastProcessedBlock = entities.Block{
 			VegaTime: time.Time{},
-			Height:   -1,
```
the problem here is if its anything other than 0 and data node has no hsitory it won't work.  it will blow on the check below:  Handling non-0 first block should be a seperate PR IMO that should not hold up this, there are other implications.



`if beginBlock.BlockNr() > lastProcessedBlock.Height+1 {
					return nil, fmt.Errorf("block height on begin block, %d, is too high, the height of the last processed block is %d",
						beginBlock.BlockNr(), lastProcessedBlock.Height)
				}
`

wwestgarth commented:
Just want to bring up again a comment I left on the core side PR from before. We do actually send out events for checkpoints during `OnCommit` in core, which I think will be after the `EndBlock` event, and I wonder if this will affect datanode-snapshot hashes.

I'm thinking about the case where we take a datanode snapshot on the block where a core checkpoint happens. The checkpoint event won't be included in datanode's snapshot because we'll commit the current block without it, but will (probably?) be included in the next snapshot as it'll get bundled up and commited to the DB as part of the next block.

If I then restore from that first datanode snapshot the checkpoint event will not be in there, but then core won't send that checkpoint event because it will restore from a point where it has already sent it. So that bit of data _won't_ be in the next datanode snapshot and might give a different hash compared to a node that hasn't restored.

Its really edge-case so don't let it block this PR, but we might want to outlaw any events being sent in `OnCommit` in core.

guoguojin committed:
chore: address code review comments about slow block timer

guoguojin committed:
Merge branch 'develop' into feat/6211-handle-block-start-end-events

ettec commented on datanode/broker/sqlstore_broker.go:
```diff
@@ -214,15 +220,15 @@ func (b *SQLStoreBroker) processBlock(ctx context.Context, dbContext context.Con
-				timeUpdate := e.(entities.TimeUpdateEvent)
 
+			switch e.Type() {
+			case events.EndBlockEvent:
```
Can we have a check here to ensure that after end block event, the next event it a being block, anything else should cause a panic, reason is it will potentially result in inconsistent state across nodes (considering restore from snapshot)

guoguojin commented:
I've created this issue for the events that occur between EndBlock/BeginBlock https://github.com/vegaprotocol/vega/issues/6414

guoguojin committed:
chore: address issue 6414

- EndBlock event is now sent OnCommit by core
- Data node will panic if an event other than BeginBlock is received
  after EndBlock

guoguojin committed:
Merge branch 'develop' into feat/6211-handle-block-start-end-events

guoguojin committed:
chore: fix linting issues

jeremyletang committed:
Merge branch 'develop' into feat/6211-handle-block-start-end-events
